File: Updog.Client/src/comment/common/comment.ts
import { User } from '@/user/common/user';
import { UserEntity } from '@/core/common/user-entity';

/**
 * A text comment attached to a post.
 */
export class Comment extends UserEntity {
    /**
     * The maximum number of characters allowed in the comment body.
     */
    public static BODY_MAX_LENGTH = 10_000;

    /**
     * The number of comments per page.
     */
    public static PAGE_SIZE = 10;

    /**
     *
     * @param id The ID of the comment.
     * @param user The user who posted it.
     * @param body The body (text) of the comment.
     * @param creationDate The date of commenting.
     * @param wasUpdated If the comment was updated.
     * @param wasDeleted If the comment was deleted.
     * @param children The children (nested) comments.
     */
    constructor(
        public id: number,
        public user: User,
        public body: string,
        public creationDate: Date,
        public wasUpdated: boolean,
        public wasDeleted: boolean,
        public children: Comment[]
    ) {
        super();
    }
}

File: Updog.Domain/Comment/Entities/Comment.cs
using System;
using System.Collections.Generic;

namespace Updog.Domain {
    /// <summary>
    /// A comment associated with a post.
    /// </summary>
    public sealed class Comment : Entity<Comment>, IUserEntity, IVotableEntity, IAuditableEntity, IUpdatable<CommentUpdate>, IDeletable {
        #region Constants
        /// <summary>
        /// The maximum number of characters in a comment body.
        /// </summary>
        public const int BodyMaxLength = 10_000;

        /// <summary>
        /// Number of comments to get per page.
        /// </summary>
        public const int PageSize = 500;
        #endregion

        #region Properties
        public int UserId { get; }
        public VoteStats Votes { get; }
        VotableEntityType IVotableEntity.VotableEntityType => VotableEntityType.Comment;
        public int PostId { get; }
        public int ParentId { get; }
        public string Body { get; private set; }
        public DateTime CreationDate { get; }
        public bool WasUpdated { get; private set; }
        public bool WasDeleted { get; private set; }
        #endregion

        #region Constructor(s)
        internal Comment(CommentCreate creationData, User user) {
            PostId = creationData.PostId;
            Body = creationData.Body;
            ParentId = creationData.ParentId;
            CreationDate = DateTime.UtcNow;
            UserId = user.Id;
            Votes = new VoteStats();
        }

        internal Comment(int id, int userId, int postId, int parentId, string body, VoteStats votes, DateTime creationDate, bool wasUpdated, bool wasDeleted) {
            Id = id;
            UserId = userId;
            PostId = postId;
            ParentId = parentId;
            Body = body;
            Votes = votes;
            CreationDate = creationDate;
            WasUpdated = wasUpdated;
            WasDeleted = wasDeleted;
        }
        #endregion

        #region Publics
        public void Update(CommentUpdate update) {
            if (WasDeleted) {
                throw new InvalidOperationException();
            }

            WasUpdated = true;
            Body = update.Body;
        }

        public void Delete() {
            this.WasDeleted = true;
            Body = "[deleted]";
        }
        #endregion
    }
}